src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,250 @@
+    }
+    return pieces;
+}
+
+ss::future<> inventory_consumer::process_path(ss::sstring path) {
+    if (auto maybe_ntp = ntp_from_path(path);
+        _ntps.contains(maybe_ntp.value())) {
+        const auto& ntp = maybe_ntp.value();
+        auto hash = xxhash_64(path.data(), path.size());
+
+        if (!_path_hashes.contains(ntp)) {
+            _path_hashes.insert({ntp, fragmented_vector<uint64_t>{}});
+        }
+
+        _path_hashes[ntp].push_back(hash);
+        _total_size += sizeof(hash);
+
+        if (_total_size >= _max_hash_size_in_memory) {
+            co_await flush();
+        }
+    }
+    co_return;
+}
+
+ss::future<>
+inventory_consumer::flush(inventory_consumer::write_all_t write_all_entries) {
+    vlog(
+      cst_log.trace,
+      "Flushing to disk, write_all_entries: {}",
+      write_all_entries);
+
+    _num_flushes += 1;
+    if (write_all_entries) {
+        for (auto& [ntp, hashes] : _path_hashes) {
+            if (!hashes.empty()) {
+                co_await flush_ntp_hashes(ntp, std::move(hashes));
+            }
+        }
+        co_return;
+    }
+
+    fragmented_vector<flush_entry> entries;
+    ranges::transform(
+      std::make_move_iterator(std::begin(_path_hashes)),
+      std::make_move_iterator(std::end(_path_hashes)),
+      std::back_inserter(entries),
+      [](auto e) {
+          return flush_entry{std::move(e.first), std::move(e.second)};
+      });
+
+    ranges::sort(entries, ranges::greater(), [](const auto& e) {
+        return e.hashes.size();
+    });
+
+    const auto target_size = _max_hash_size_in_memory / 3;
+    vlog(
+      cst_log.trace,
+      "Before flush, size in memory {}, target {}",
+      _total_size,
+      target_size);
+
+    size_t idx = 0;
+    while (idx < entries.size() && _total_size >= target_size) {
+        _total_size -= sizeof(entries[idx].hashes[0])
+                       * entries[idx].hashes.size();
+
+        const auto path = entries[idx].ntp.path();
+        co_await flush_ntp_hashes(
+          std::move(entries[idx].ntp), std::move(entries[idx].hashes));
+        vlog(
+          cst_log.trace,
+          "Flushed ntp {}, size in memory {}, target {}",
+          path,
+          _total_size,
+          target_size);
+        idx += 1;
+    }
+
+    vlog(cst_log.trace, "After flush, size in memory {}", _total_size);
+    for (; idx < entries.size(); ++idx) {
+        _path_hashes.insert(
+          {std::move(entries[idx].ntp), std::move(entries[idx].hashes)});
+    }
+}
+
+ss::future<> inventory_consumer::flush_ntp_hashes(
+  model::ntp ntp, fragmented_vector<uint64_t> hashes) {
+    if (!_streams.contains(ntp)) {
+        auto ntp_hash_path = _hash_store_path / std::string_view{ntp.path()};
+        const auto hash_path = ntp_hash_path.string();
+
+        const auto& dir_name = ntp_hash_path.remove_filename().string();
+        if (!co_await ss::file_exists(dir_name)) {
+            co_await ss::recursive_touch_directory(dir_name);
+        }
```

Comment:
should we put all these files into the cloud cache? if the data is large, should we integrate its size with space usage?

Reply:
I avoided putting them in cache because we could lose the data on cache churn and with a busy cache the deep scrub might never run, but this could definitely cause a problem because of space usage. Although the inventory service clears out data for the previous scrub, it only will do so on the next run. At best a scrubber working with a specifc NTP could remove the hash file for that NTP after it has finished the scrub.

To give a sense of disk usage, in my local environment, for 100 million paths the file produced is around 760MiB in size, so for multiple partitions this could be a fair bit of disk usage. Also, these files do not compress well, presumably because they are largely composed of integers and not text data. 

I will also try to store the hashes in `deltafor_encoder`, perhaps it has better space usage, although IIRC it works best for sequential integers.

Reply:
IMO `deltafor_encoder` will not be able to help here because the integers are random.

Reply:
> I will also try to store the hashes in deltafor_encoder, perhaps it has better space usage, although IIRC it works best for sequential integers.

The deltafor encoder takes almost the exact same space in memory as the fragmented vector - 760~ MiB.

Reply:
I tried with a billion hashes just to test, and additionally sorted the hashes before adding them to `deltafor_encoder` - it is quite smaller than a plain fragmented vector, although still very large. The  `deltafor_encoder` size is around 5GiB and the fragmented vector is around 8GiB in size.